src/librustc_mir_build/hair/pattern/check_match.rs

```diff
@@ -612,81 +617,109 @@ fn check_legality_of_move_bindings(cx: &mut MatchVisitor<'_, '_>, has_guard: boo
-            for sp in conflicts_mut_ref {
-                err.span_label(sp, &format!("{} borrow occurs here", also));
-            }
-            err.emit();
+    // We now have `ref $mut_outer binding @ sub` (semantically).
+    // Recurse into each binding in `sub` and find mutability or move conflicts.
+    let mut conflicts_move = Vec::new();
+    let mut conflicts_mut_mut = Vec::new();
+    let mut conflicts_mut_ref = Vec::new();
+    sub.each_binding(|_, hir_id, span, _| match tables.extract_binding_mode(sess, hir_id, span) {
+        Some(ty::BindByReference(mut_inner)) => match (mut_outer, mut_inner) {
+            (Mutability::Not, Mutability::Not) => {} // Both sides are `ref`.
+            (Mutability::Mut, Mutability::Mut) => conflicts_mut_mut.push(span), // 2x `ref mut`.
+            _ => conflicts_mut_ref.push(span),       // `ref` + `ref mut` in either direction.
+        },
+        Some(ty::BindByValue(_)) if is_binding_by_move(cx, hir_id, span) => {
+            conflicts_move.push(span) // `ref mut?` + by-move conflict.
         }
+        Some(ty::BindByValue(_)) | None => {} // `ref mut?` + by-copy is fine.
     });
+
+    // Report errors if any.
+    if !conflicts_mut_mut.is_empty() {
+        // Report mutability conflicts for e.g. `ref mut x @ Some(ref mut y)`.
+        let msg = &format!("cannot borrow `{}` as mutable more than once at a time", name);
+        sess.struct_span_err(pat.span, msg)
+            .span_label(binding_span, "first mutable borrow occurs here")
+            .span_labels(conflicts_mut_mut, "another mutable borrow occurs here")
+            .span_labels(conflicts_mut_ref, "also borrowed as immutable here")
+            .span_labels(conflicts_move, "also moved here")
+            .emit();
+    } else if !conflicts_mut_ref.is_empty() {
+        // Report mutability conflicts for e.g. `ref x @ Some(ref mut y)` or the converse.
+        let (primary, also) = match mut_outer {
+            Mutability::Mut => ("mutable", "immutable"),
+            Mutability::Not => ("immutable", "mutable"),
+        };
+        let msg = &format!(
+            "cannot borrow `{}` as {} because it is also borrowed as {}",
+            name, also, primary,
+        );
+        sess.struct_span_err(pat.span, msg)
+            .span_label(binding_span, format!("{} borrow occurs here", primary))
+            .span_labels(conflicts_mut_ref, format!("{} borrow occurs here", also))
+            .span_labels(conflicts_move, "also moved here")
+            .emit();
+    } else if !conflicts_move.is_empty() {
+        // Report by-ref and by-move conflicts, e.g. `ref x @ y`.
+        let msg = &format!("cannot move out of `{}` because it is borrowed", name);
+        sess.struct_span_err(pat.span, msg)
+            .span_label(binding_span, format!("borrow of `{}` occurs here", name))
+            .span_labels(conflicts_move, format!("move out of `{}` occurs here", name))
```

Comment:
(1) ...and by:
> This applies for all the errors here.

...you mean that this should say `"value moved here" as well as not saying "out of `{}`" in the message?
(To say "value moved into `{}`" we would have to store the names of the by-value sub-bindings, which seems fine; i'll fix that.)

Reply:
yes